pkgs/development/python-modules/pygobject/3.nix

```diff
@@ -42,15 +42,15 @@ buildPythonPackage rec {
   ];
 
   buildInputs = [
-    # # .so files link to this
+    cairo
+    # .so files link to this
     glib
   ] ++ lib.optionals stdenv.isDarwin [
     ncurses
   ];
 
   propagatedBuildInputs = [
     pycairo
-    cairo
```

Comment:
Are you sure this will not break stuff like pkg-config lookup? Or whatever was the reason for this propagation.

Reply:
pycairo itself should propagate  cairo
```diff
diff --git a/pkgs/applications/version-management/meld/default.nix b/pkgs/applications/version-management/meld/default.nix
index 224efe8c1ac2..ed7da735a585 100644
--- a/pkgs/applications/version-management/meld/default.nix
+++ b/pkgs/applications/version-management/meld/default.nix
@@ -45,11 +45,12 @@ python3.pkgs.buildPythonApplication rec {
     gtksourceview4
     gsettings-desktop-schemas
     gnome.adwaita-icon-theme
+    python3.pkgs.pygobject3
   ];

   propagatedBuildInputs = with python3.pkgs; [
-    pygobject3
-    pycairo
+    pygobject3.out
+    pycairo.out
   ];

   postPatch = ''
diff --git a/pkgs/development/python-modules/pycairo/default.nix b/pkgs/development/python-modules/pycairo/default.nix
index 84c64b79df06..76e3e6a2faac 100644
--- a/pkgs/development/python-modules/pycairo/default.nix
+++ b/pkgs/development/python-modules/pycairo/default.nix
@@ -32,12 +32,15 @@ buildPythonPackage rec {
     pkg-config
   ];

-  buildInputs = [
-    cairo
-  ] ++ lib.optionals (pythonOlder "3.9") [
+  buildInputs = lib.optionals (pythonOlder "3.9") [
     libxcrypt
   ];

+  propagatedBuildInputs = [
+    # the header includes cairo.h
+    cairo
+  ];
+
   nativeCheckInputs = [
     pytestCheckHook
   ];
diff --git a/pkgs/development/python-modules/pygobject/3.nix b/pkgs/development/python-modules/pygobject/3.nix
index b8337aea5621..fe70790d3c59 100644
--- a/pkgs/development/python-modules/pygobject/3.nix
+++ b/pkgs/development/python-modules/pygobject/3.nix
@@ -50,7 +50,6 @@ buildPythonPackage rec {

   propagatedBuildInputs = [
     pycairo
-    cairo
   ];

   mesonFlags = [
```

the cairo propagation in pygobject was added in https://github.com/NixOS/nixpkgs/pull/5893

Reply:
> Are you sure this will not break stuff like pkg-config lookup?

No


> Or whatever was the reason for this propagation.

I think it was a mistake from 8 years ago.

Reply:
Looking at the outputs some more, I think removal of cairo from propagation is correct:

```ShellSession
$ rg -i cairo result-dev/

$ rg -i cairo result/
result/lib/python3.11/site-packages/pygtkcompat/pygtkcompat.py
172:    # pangocairo
173:    gi.require_version('PangoCairo', '1.0')
174:    from gi.repository import PangoCairo
####    ⬆️⬆️⬆️⬆️⬆️ This is part of Pango
175:    _patch_module('pangocairo', PangoCairo)

result/lib/python3.11/site-packages/gi/overrides/Gdk.py
161:        from gi.repository import cairo as _cairo
####    ⬆️⬆️⬆️⬆️⬆️ This is part of gobject-introspection or maybe _gi_cairo.cpython-311-x86_64-linux-gnu.so
162:        Rectangle = _cairo.RectangleInt
176:        def cairo_create(self):
177:            return Gdk.cairo_create(self)
191:        def cairo_create(self):
192:            return Gdk.cairo_create(self)

result/lib/python3.11/site-packages/gi/__init__.py
160:        Introspection namespace of the foreign module (e.g. "cairo")
171:        import cairo
172:        gi.require_foreign('cairo')
####    ⬆️⬆️⬆️⬆️⬆️ This is just docs
```

It should still remain in `buildInputs` to be able to build `_gi_cairo.cpython-311-x86_64-linux-gnu.so`.

Reply:
I think @Artturin’s suggestion to propagate `cairo` from `pycairo` is correct as well (`cairo` is in `Requires` section of `py3cairo.pc`’), in which case the `.out` hack will be needed. But that is functionally orthogonal to this change (even though it will basically revert it – but we cannot do much about it without https://github.com/NixOS/nixpkgs/pull/102613)

Reply:
So we can merge this as is?

Reply:
Yeah, I think this is a net improvement.